### ISSUE TITLE
`less.render` callback does not print any syntax or other errors

### ISSUE THREAD

**Issue by kwketh** [less/less.js#2546]
Hi all.

I have observed an odd behaviour in `less.render` callback where program execution stops on any error in the callback scope but it does not print any trace or any error whatsoever.

I have been unable to trace the issue, I am unsure if this is language semantics (use of `try` and `catch` statements), if it is node.js issue or the less.js library. I understand there are workarounds and possible solutions but regardless I don't believe this expected behaviour. The behaviour is more odd that it depends on the less contents provided. The callback (in both scenarios attached) is provided with a 'truthy' error from `less.js` but **the callback scope behaves differently** (one stops execution, the other does not).

Can anyone see what's going on wrong? I have attached two `.js` one that works `test-working.js` and one that reproduces the issue `test-failing.js`.

Forgive me if I have missed something obvious, it has been while since I used node. 

Code for `test-working.js`: https://gist.github.com/kwketh/b64bbf3994445e2b9c27
**Output:**
```
$ node test-working.js
callback works as expected, error will be thrown.

/Users/kwketh/[ ... ]/test-working.js:6
    callUndefinedFunction()
    ^
ReferenceError: callUndefinedFunction is not defined
...
```

Code for `test-failing.js`: https://gist.github.com/kwketh/4c3c47308d619a640db1
**Output:**
```
$ node test-failing.js
callback works but does not display any errors.
```

Many thanks.

**Comment by seven-phases-max** [comment]
I would simply suggest to change your code to (if you need a normal "sync" behaviour with no internal Less exception handlers interfering):
```
var output;
less.render(input, options, function(result) {
    output = result;
});
    
console.log("do what you normally do here");
```
(note this is just a suggestion I do not imply if it's a bug or not a bug or something).

**Comment by kwketh** [comment]
@seven-phases-max 
I have tried many different solutions as such and I have not managed to get "sync" behaviour working. I have tried `sync: true` and `async: false` in the options and followed all the docs correctly, the final `console.log` would just print `undefined` (since the render call is asynchronous).

How do you enable the "sync" behaviour in node.js?

**Comment by seven-phases-max** [comment]
Hmm, this `render` form is actually synchronous (unless there's some bug in the version you're using). E.g. to make sure try:
```
var output;
less.render(<some-vaery-big-less-input-here>, options, function(e, result) {
    console.log("done.");
});

console.log("after return");
```

"done." should never go before "after return".

**Comment by kwketh** [comment]
You are right @seven-phases-max, I have no idea how I missed that. This is definitely a good solution. Thanks a lot, I really appreciate it.

On the other note, I suggest to keep the issue open as it is rather major that a syntax or reference error could propagate down and lead to an unexpected behaviour in a program without any trace or information. It has took me two hours to trace down what was going wrong with my application and why no errors are thrown or why the execution stopped in the callback without any information. If this is actual bug and can be fixed then perhaps can save a lot of hours for somebody in the future.

**Comment by kwketh** [comment]
@seven-phases-max: It works perfectly with a less contents as you described `<some-very-big-less-input-here>`, however I just noticed that as soon as I add `@import "file.less"` on the first line it seems that `less.render` is no longer synchronous and the `console.log("done")` prints after `"after return"`. 

Is this expected behaviour?

**Comment by seven-phases-max** [comment]
>that as soon as I add  @import "file.less"  on the first line it seems that  less.render  is no longer synchronous and the  console.log("done")  prints after  "after return" .

Indeed, you're right. My apologizes for misdirecting. (Honestly I did no look too deep for the changes at that part so I have no idea if it's expcted/intended or not).

Then for your initial problem another (just a bit more verbose) solution would be:
```js
    less.render("@import 'less-own-benchmark.less';", options, function(e, result) {
        // just set you own EH to catch anything wrong before the compiler does
        try {
            console.log("callback works but does not display any errors.");
            callUndefinedFunction();
        }
        catch (e) {
            // console.log(e);
            console.log("execution stopped, no errors in either stdout or stderr.");
            // throw(e); // rethrow if necessary
        }
    });
```

**Comment by kwketh** [comment]
You are right, this would work. I just don't think many people actually wrap their whole callback like ever in a try statement to catch a syntax error, it wouldn't seem obvious that you have to do this for all callbacks. Ideally whoever is catching it, I believe `less.js` does somewhere, it should at least print th message to `stderr` via `process.stdout.write` and/or terminate the process with error code 1 as opposed to no output and everything seeming to succeed (error code 0).

Regarding the other solution using synchronous `less.render`, I have found solution (in very old issue #615) where there is an undocumented option `syncImport` and if set to `true` then my the `less.parse` call is synchronous even if `@import` statement is used.

Both solutions seem to be a good workaround (using either "sync" or "try-catch"), but I suggest to keep the bug open.

**Comment by publicocean0** [comment]
i tried it ... but dont work 
var done=false;
less.render(input, options, function(err, result) {
    // just set you own EH to catch anything wrong before the compiler does:
   done=true;
    try {
        console.log("callback works");
        callUndefinedFunction();
    }
    catch (e) {
        // console.log(e);
        console.log("execution stopped");
        // throw(e); // rethrow if necessary
    }
});
while (!done) ;

**Comment by publicocean0** [comment]
ops i didnt read all ahaha ... using syncImport it works :)

**Comment by publicocean0** [comment]
i suggested to add this flag in documentation ... it is a common problem for many developers i saw ....:) ciao
